### PR TITLE
llvm backend pretty printing tool

### PR DESCRIPTION
**Pull request by dwightguth**
This is the llvm backend part of a pair of PRs used to support pretty-printing kore terms with a vastly reduced overhead. Some logic is not implemented yet, but it is sufficient to unparse some simple IMP programs. I will add more features as I go if this proves useful to people.

Second PR is here: https://github.com/kframework/k/pull/1211

### CODE REVIEW

File: lib/ast/AST.cpp [hunk 1]
```diff
@@ -285,6 +286,512 @@ sptr<KOREPattern> KORECompositePattern::expandAliases(KOREDefinition *def) {
...
+            indent++;
+            localIndent++;
+            break;
+          case 'd':
+            indent--;
+            localIndent--;
+            break;
+          case 'c':
+            if (data.colors.count(name)) {
+              color(out, data.colors.at(name)[localColor++]);
```
Comment (virgil-serbanuta) [inline lib/ast/AST.cpp:611]: I don't understand this. Why is `localColor` incremented here? Could it happen that it is incremented past the end of that vector?

File: lib/ast/AST.cpp [hunk 2]
```diff
@@ -285,6 +286,512 @@ sptr<KOREPattern> KORECompositePattern::expandAliases(KOREDefinition *def) {
...
+            }
+            break;
+          case 'r':
+            if (hasColor) {
+              append(out, "\x1b[0m");
```
Comment (virgil-serbanuta) [inline lib/ast/AST.cpp:616]: Could you add a constant for this (e.g. `RESET`)?

File: tools/kprint/main.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  std::map<std::string, std::string> formats;
+  formats["kseq"] = "%1 ~> %2";
+  formats["dotk"] = ".";
+  formats["inj"] = "%1";
+  std::map<std::string, std::string> hooks;
+  std::set<std::string> assocs;
+  std::set<std::string> comms;
+  std::map<std::string, std::vector<std::string>> colors;
+
+  std::ifstream iformats(argv[1] + std::string("/format-att.txt"));
```
Comment (virgil-serbanuta) [inline tools/kprint/main.cpp:36]: Who is supposed to write these files? The frontend, the user, or something else? I think I could infer what most of them mean, but, although I could follow what happens with that data, I don't think I actually understand `color-att.txt`.

File: bin/llvm-krun [hunk 4]
```diff
@@ -3,13 +3,16 @@ set -e
```
Comment (virgil-serbanuta) [inline bin/llvm-krun:3]: Could you give a command line example for this script?

File: lib/ast/AST.cpp (hunk 1 above)
Comment (dwightguth) [reply lib/ast/AST.cpp:619]: That would only happen if the user incorrectly specifies their `format` or `colors` attributes.

File: tools/kprint/main.cpp (hunk 3 above)
Comment (dwightguth) [reply tools/kprint/main.cpp:37]: It is generated by the frontend; see the accompanying PR.

Comment (virgil-serbanuta) [reply tools/kprint/main.cpp:37]: Ok, but why does it make sense to have multiple colors for the same thing, and then to use each of them once?

Comment (dwightguth) [reply tools/kprint/main.cpp:37]: well, you may want to color different parts of the same production differently. consider for example the following production:
```
syntax Stmt ::= "if" "(" Stmt ")" Block "else" Block
```

Presumably you may want to color the parentheses differently than the keywords.

**Pull request merged**